### DISCUSSION TITLE
Presence shape for virtual character controllers

### DISCUSSION THREAD

**Discussion by jankrassnigg** [jrouwe/JoltPhysics#239] [Q&A]
I want my character to interact with world (trigger sensors, be hit by raycasts, etc). As I see it, there are two options:

1. Have an additional kinematic shape that I move around with the CC. This is what I first tried, and this is why I ran into issue #238.
2. Add a JPH::Character in the same place as my JPH::VirtualCharacter

I guess both would work, but I am wondering, are there any reasons why a JPH::Character would be beneficial? I guess I would make the 'presence shape' (be it kinematic or a Character) a bit smaller than the VirtualCharacter, but I fear that I may still encounter situations where moving the JPH::Character to the desired location may not work (because of sliding and being able to get stuck), whereas using a kinematic shape is guaranteed to work.

So, mainly curious, you mentioned before that the JPH::Character is meant for this task, but it feels like a bit overkill. Also JPH::Character and JPH::VirtualCharacter have a lot of overlap in functionality (ground contact and so on), so it doesn't really feel like JPH::Character's real purpose is that well defined.

**Comment by jrouwe** [comment]
From the top of my head, this is what we do:

1. AI characters use a Character. It is configured to only collide with dynamic objects and has no gravity so it effectively moves however the AI wants it to move (over the navmesh) and is only there to push dynamic things away. The Character class was made for when complex sliding logic / stair walking is not required. It is cheap to execute compared to CharacterVirtual. If you don't have a navmesh, you could use the Character like it is used in the Samples application (with gravity and colliding with the static environment) to get cheap collision detection for AI.

2. The player uses a CharacterVirtual for its main movement and collision detection. We have an inner Character that is a little bit smaller than the CharacterVirtual and again configured not to collide with the static world and has no gravity. It follows the CharacterVirtual by setting the velocity (keyframed fashion, except that the mass is not infinite) and the only purpose of it is to give the player a presence in the physics world (against e.g. ray casts or very fast moving physics objects that are not picked up when CharacterVirtual moves). Most of the time CharacterVirtual is responsible for pushing things away, but if the penetrations are deep enough then Character will help push things out.

Character is just a thin wrapper around a rigid body (its main purpose is to setup the inertia in such a way that the capsule doesn't rotate). The overhead of determining if you're on the ground or not occurs only if you call Character::PostSimulation (which we don't for the Character used by the CharacterVirtual). I think using a normal Body would be an equally valid solution. So I think your solutions 1 and 2 are both good.
